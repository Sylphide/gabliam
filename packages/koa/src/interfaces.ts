import { koaRouter } from './koa';

/**
 * Config for koa plugin
 */
export interface KoaPluginConfig {
  /**
   * Root path of koa plugin
   */
  rootPath: string;

  /**
   * Port of koa
   */
  port: number;

  /**
   * Hostname of koa
   */
  hostname: string;
}

/**
 * Represent a method that create an koa router
 */
export type RouterCreator = (path?: string) => koaRouter;
