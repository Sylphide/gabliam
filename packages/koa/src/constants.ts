/** Type in registry */
export const TYPE = {
  Controller: 'ControllerType'
};

export const METADATA_KEY = {
  controller: '_controller',
  controllerMethod: '_controller-method',
  controllerParameter: '_controller-parameter',
  middleware: '_middleware',
  MiddlewareConfig: '_middlewareConfig'
};

export enum PARAMETER_TYPE {
  REQUEST,
  RESPONSE,
  PARAMS,
  QUERY,
  BODY,
  HEADERS,
  COOKIES,
  NEXT
}

export const DEFAULT_ROUTING_ROOT_PATH = '/';

export const KOA_PLUGIN_CONFIG = Symbol('GABLIAM/KOA_PLUGIN_CONFIG');

export const APP = Symbol('GABLIAM/KOA_APP');

export const DEFAULT_PARAM_VALUE = '**$$@gabliam/koa:DEFAULT_PARAM_VALUE$$**';

/**
 * Represent a custom router creator
 *
 * If you want a custom router use this const for create your router
 *
 * ## Sample
 * @Config()
 * class Conf {
 *  @Bean(CUSTOM_ROUTER_CREATOR)
 *  custom() {
 *    return () =>
 *      e.Router({
 *        caseSensitive: true
 *      });
 *  }
 * }
 */
export const CUSTOM_ROUTER_CREATOR = Symbol(
  'GABLIAM/CUSTOM_KOA_ROUTER_CREATOR'
);

export const SERVER = Symbol('GABLIAM/KOA_SERVER');

export const ERRORS_MSGS = {
  DUPLICATED_CONFIG_DECORATOR: `Cannot apply @KoaConfig decorator multiple times on same method.`,
  DUPLICATED_CONTROLLER_DECORATOR: `Cannot apply @Controller and @RestController decorator multiple times.`
};
