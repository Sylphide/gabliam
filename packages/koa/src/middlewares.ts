import { KoaConfiguration } from './interfaces';

/**
 * Middleware Config
 *
 * This class is a storage of all middlewares
 */
export class MiddlewareConfig {
  private _middlewares: KoaConfiguration[] = [];

  addMiddleware(middleware: KoaConfiguration) {
    this._middlewares.push(middleware);
  }

  get middlewares() {
    return this._middlewares.slice();
  }
}
