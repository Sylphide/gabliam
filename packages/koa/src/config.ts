import { PluginConfig, Bean, Value, Joi } from '@gabliam/core';
import { DEFAULT_ROUTING_ROOT_PATH, KOA_PLUGIN_CONFIG } from './constants';
import { KoaPluginConfig } from './interfaces';

@PluginConfig()
export class RestPluginConfig {
  @Value('application.koa.rootPath', Joi.string())
  rootPath = DEFAULT_ROUTING_ROOT_PATH;

  @Value('application.koa.port', Joi.number().positive())
  port: number = process.env.PORT ? parseInt(process.env.PORT!, 10) : 3000;

  @Value('application.koa.hostname', Joi.string())
  hostname!: string;

  @Bean(KOA_PLUGIN_CONFIG)
  restConfig(): KoaPluginConfig {
    return {
      rootPath: this.rootPath,
      port: this.port,
      hostname: this.hostname
    };
  }
}
