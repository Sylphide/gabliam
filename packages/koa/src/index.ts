import { APP, SERVER } from './constants';
import { KoaPlugin } from './koa-plugin';

export * from '@gabliam/web-core';

export * from './middlewares';
export * from './koa';
export * from './interfaces';
export * from './decorators';
export * from 'http-status-codes';
export { APP, SERVER, KoaPlugin as default };
