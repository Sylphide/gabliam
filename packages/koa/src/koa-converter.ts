import {
  Container,
  InjectContainer,
  INJECT_CONTAINER_KEY,
  Service,
  toPromise,
} from '@gabliam/core';
import {
  createInterceptorResolver,
  extractParameters,
  getParameterMetadata,
  isInterceptor,
  getContext,
} from '@gabliam/web-core';
import { koa } from './koa';

/**
 * Class for convert interceptor to Koa Middleware
 */
@InjectContainer()
@Service()
export class KoaConverter {
  public interceptorToMiddleware(clazz: any): koa.Middleware {
    const container = <Container>(this as any)[INJECT_CONTAINER_KEY];
    const instance = createInterceptorResolver(container)(clazz);
    if (!isInterceptor(instance)) {
      throw new Error();
    }

    return async (context: koa.Context, next: () => Promise<any>) => {
      const args = extractParameters(
        instance,
        'intercept',
        null,
        getContext(context.req),
        next,
        getParameterMetadata(instance, 'intercept')
      );
      await toPromise((instance['intercept'] as any)(...args));
      await next();
    };
  }
}
