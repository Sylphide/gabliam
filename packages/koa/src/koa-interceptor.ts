import {
  AfterResponseInterceptor,
  Interceptor,
  InterceptorMethod,
  InterceptorConstructor,
} from '@gabliam/web-core';
import { METADATA_KEY } from './constants';
import { koaRouter } from './koa';
import { Service } from '@gabliam/core';

/**
 * Test if target is an koa interceptor
 * @param target any
 */
export const isKoaInterceptor = (target: any) =>
  Reflect.getOwnMetadata(
    METADATA_KEY.koaInterceptor,
    target.constructor || target
  ) === true;

/**
 * KoaInterceptor decorator
 * class is an koa interceptor.
 * Must return koaRouter middleware
 */
const KoaInterceptor = () => (target: any) => {
  Reflect.defineMetadata(METADATA_KEY.koaInterceptor, true, target);
};

/**
 * Convert a Koa router middleware to an express interceptor
 */
export const toInterceptor = (
  mid: koaRouter.IMiddleware,
  type: InterceptorMethod = 'intercept'
): InterceptorConstructor => {
  let clazz: InterceptorConstructor;
  if (type === 'intercept') {
    clazz = class implements Interceptor {
      intercept() {
        return mid;
      }
    };
  } else {
    clazz = class implements AfterResponseInterceptor {
      afterResponse() {
        return mid;
      }
    };
  }
  Service()(clazz);
  KoaInterceptor()(clazz);
  return clazz;
};
