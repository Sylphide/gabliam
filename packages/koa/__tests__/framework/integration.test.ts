import {
  Controller,
  RestController,
  Get,
  Post,
  Put,
  Patch,
  Head,
  Delete,
  Method,
  koaRouter
} from '../../src/index';
import { KoaPluginTest } from '../koa-plugin-test';
import { Config, Bean } from '@gabliam/core';
import { CUSTOM_ROUTER_CREATOR } from '../../src/constants';

let appTest: KoaPluginTest;

beforeEach(async () => {
  appTest = new KoaPluginTest();
});

afterEach(async () => {
  await appTest.destroy();
});

describe('Integration Tests:', () => {
  [Controller, RestController].forEach(decorator => {
    describe(`decorator ${decorator.name}`, () => {
      test('should work for async controller methods', async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          async getTest() {
            return new Promise(resolve => {
              setTimeout(resolve, 100, 'GET');
            });
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const response = await appTest
          .supertest()
          .get('/')
          .expect(200);
        expect(response).toMatchSnapshot();
      });

      test('should work for async controller methods that fails', async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          public getTest() {
            return new Promise((resolve, reject) => {
              setTimeout(reject, 100, 'GET');
            });
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const response = await appTest
          .supertest()
          .get('/')
          .expect(500);
        expect(response).toMatchSnapshot();
      });

      test('should work for methods which call nextFunc()', async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          public async getTest(
            ctx: koaRouter.IRouterContext,
            nextFunc: () => Promise<any>
          ) {
            await nextFunc();
          }

          @Get('/')
          public getTest2(ctx: koaRouter.IRouterContext) {
            return 'GET';
          }
        }

        appTest.addClass(TestController);
        await appTest.build();
        const response = await appTest
          .supertest()
          .get('/')
          .expect(200);
        expect(response).toMatchSnapshot();
      });

      test('should work for async methods which call nextFunc()', async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          public getTest(
            ctx: koaRouter.IRouterContext,
            nextFunc: () => Promise<any>
          ) {
            return new Promise(resolve => {
              setTimeout(
                () => {
                  resolve(nextFunc());
                },
                100,
                'GET'
              );
            });
          }

          @Get('/')
          public getTest2(ctx: koaRouter.IRouterContext) {
            return 'GET';
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const response = await appTest
          .supertest()
          .get('/')
          .expect(200);
        expect(response).toMatchSnapshot();
      });

      test('should work for async methods called by nextFunc()', async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          public async getTest(
            ctx: koaRouter.IRouterContext,
            nextFunc: () => Promise<any>
          ) {
            await nextFunc();
          }

          @Get('/')
          public getTest2(ctx: koaRouter.IRouterContext) {
            return new Promise(resolve => {
              setTimeout(resolve, 100, 'GET');
            });
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const response = await appTest
          .supertest()
          .get('/')
          .expect(200);
        expect(response).toMatchSnapshot();
      });

      test('should work for each shortcut decorator', async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          public getTest(ctx: koaRouter.IRouterContext) {
            ctx.body = 'GET';
          }
          @Post('/')
          public postTest(ctx: koaRouter.IRouterContext) {
            ctx.body = 'POST';
          }
          @Put('/')
          public putTest(ctx: koaRouter.IRouterContext) {
            ctx.body = 'PUT';
          }
          @Patch('/')
          public patchTest(ctx: koaRouter.IRouterContext) {
            ctx.body = 'PATCH';
          }
          @Head('/')
          public headTest(ctx: koaRouter.IRouterContext) {
            ctx.body = 'HEAD';
          }
          @Delete('/')
          public deleteTest(ctx: koaRouter.IRouterContext) {
            ctx.body = 'DELETE';
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const agent = appTest.supertest();

        const rd = await agent.delete('/').expect(200);
        expect(rd).toMatchSnapshot();

        const rh = await agent.head('/').expect(200);
        expect(rh).toMatchSnapshot();

        const rpatch = await agent.patch('/').expect(200);
        expect(rpatch).toMatchSnapshot();

        const rput = await agent.put('/').expect(200);
        expect(rput).toMatchSnapshot();

        const rpost = await agent.post('/').expect(200);
        expect(rpost).toMatchSnapshot();

        const rget = await agent.get('/').expect(200);
        expect(rget).toMatchSnapshot();
      });

      test('should work for more obscure HTTP methods using the httpMethod decorator', async () => {
        @decorator('/')
        class TestController {
          @Method('propfind', '/')
          public getTest(ctx: koaRouter.IRouterContext) {
            ctx.body = 'PROPFIND';
          }
        }

        appTest.addClass(TestController);
        await appTest.build();

        const response = await appTest
          .supertest()
          .propfind('/')
          .expect(200);
        expect(response).toMatchSnapshot();
      });

      test('should use returned values as response', async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          public getTest(ctx: koaRouter.IRouterContext) {
            return { hello: 'world' };
          }
        }

        appTest.addClass(TestController);
        await appTest.build();

        const response = await appTest
          .supertest()
          .get('/')
          .expect(200);
        expect(response).toMatchSnapshot();
      });

      test('should use custom router passed from configuration', async () => {
        @decorator('/CaseSensitive')
        class TestController {
          @Get('/Endpoint')
          public get() {
            return 'Such Text';
          }
        }

        @Config()
        class Conf {
          @Bean(CUSTOM_ROUTER_CREATOR)
          custom() {
            return (prefix: string) =>
              new koaRouter({
                prefix,
                sensitive: true
              });
          }
        }
        appTest.addClass(Conf);
        appTest.addClass(TestController);
        await appTest.build();

        const agent = appTest.supertest();

        const expectedSuccess = await agent
          .get('/CaseSensitive/Endpoint')
          .expect(200);

        expect(expectedSuccess).toMatchSnapshot();

        const expectedNotFound1 = await agent
          .get('/casesensitive/endpoint')
          .expect(404);

        expect(expectedNotFound1).toMatchSnapshot();

        const expectedNotFound2 = await agent
          .get('/CaseSensitive/endpoint')
          .expect(404);

        expect(expectedNotFound2).toMatchSnapshot();
      });

      test('should use custom routing configuration', async () => {
        @decorator('/ping')
        class TestController {
          @Get('/endpoint')
          public get() {
            return 'pong';
          }
        }

        appTest.addClass(TestController);
        appTest.addConf('application.koa.rootPath', '/api/v1');
        await appTest.build();

        const response = await appTest
          .supertest()
          .get('/api/v1/ping/endpoint')
          .expect(200);

        expect(response).toMatchSnapshot();
      });
    });
  });
});
