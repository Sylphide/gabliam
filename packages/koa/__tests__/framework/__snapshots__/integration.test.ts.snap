// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration Tests: decorator createControllerDecorator should use custom router passed from configuration 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "9",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "Such Text",
}
`;

exports[`Integration Tests: decorator createControllerDecorator should use custom router passed from configuration 2`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "9",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 404,
  "text": "Not Found",
}
`;

exports[`Integration Tests: decorator createControllerDecorator should use custom router passed from configuration 3`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "9",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 404,
  "text": "Not Found",
}
`;

exports[`Integration Tests: decorator createControllerDecorator should use custom routing configuration 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "pong",
}
`;

exports[`Integration Tests: decorator createControllerDecorator should use returned values as response 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "17",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "{\\"hello\\":\\"world\\"}",
}
`;

exports[`Integration Tests: decorator createControllerDecorator should work for async controller methods 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Integration Tests: decorator createControllerDecorator should work for async controller methods that fails 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "21",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 500,
  "text": "Internal Server Error",
}
`;

exports[`Integration Tests: decorator createControllerDecorator should work for async methods called by nextFunc() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Integration Tests: decorator createControllerDecorator should work for async methods which call nextFunc() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Integration Tests: decorator createControllerDecorator should work for each shortcut decorator 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "6",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "DELETE",
  },
  "status": 200,
  "text": "DELETE",
}
`;

exports[`Integration Tests: decorator createControllerDecorator should work for each shortcut decorator 2`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "HEAD",
  },
  "status": 200,
  "text": undefined,
}
`;

exports[`Integration Tests: decorator createControllerDecorator should work for each shortcut decorator 3`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "PATCH",
  },
  "status": 200,
  "text": "PATCH",
}
`;

exports[`Integration Tests: decorator createControllerDecorator should work for each shortcut decorator 4`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "PUT",
  },
  "status": 200,
  "text": "PUT",
}
`;

exports[`Integration Tests: decorator createControllerDecorator should work for each shortcut decorator 5`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "POST",
  },
  "status": 200,
  "text": "POST",
}
`;

exports[`Integration Tests: decorator createControllerDecorator should work for each shortcut decorator 6`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Integration Tests: decorator createControllerDecorator should work for methods which call nextFunc() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Integration Tests: decorator createControllerDecorator should work for more obscure HTTP methods using the httpMethod decorator 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "8",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "PROPFIND",
  },
  "status": 200,
  "text": "PROPFIND",
}
`;

exports[`Integration Tests: decorator createControllerDecorator should work with config 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "14",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "config ok !!!!",
}
`;

exports[`Integration Tests: decorator createRestControllerDecorator should use custom router passed from configuration 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "11",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "\\"Such Text\\"",
}
`;

exports[`Integration Tests: decorator createRestControllerDecorator should use custom router passed from configuration 2`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "9",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 404,
  "text": "Not Found",
}
`;

exports[`Integration Tests: decorator createRestControllerDecorator should use custom router passed from configuration 3`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "9",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 404,
  "text": "Not Found",
}
`;

exports[`Integration Tests: decorator createRestControllerDecorator should use custom routing configuration 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "6",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "\\"pong\\"",
}
`;

exports[`Integration Tests: decorator createRestControllerDecorator should use returned values as response 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "17",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "{\\"hello\\":\\"world\\"}",
}
`;

exports[`Integration Tests: decorator createRestControllerDecorator should work for async controller methods 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "\\"GET\\"",
}
`;

exports[`Integration Tests: decorator createRestControllerDecorator should work for async controller methods that fails 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "21",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 500,
  "text": "Internal Server Error",
}
`;

exports[`Integration Tests: decorator createRestControllerDecorator should work for async methods called by nextFunc() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "\\"GET\\"",
}
`;

exports[`Integration Tests: decorator createRestControllerDecorator should work for async methods which call nextFunc() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "\\"GET\\"",
}
`;

exports[`Integration Tests: decorator createRestControllerDecorator should work for each shortcut decorator 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "6",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "DELETE",
  },
  "status": 200,
  "text": "DELETE",
}
`;

exports[`Integration Tests: decorator createRestControllerDecorator should work for each shortcut decorator 2`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "HEAD",
  },
  "status": 200,
  "text": undefined,
}
`;

exports[`Integration Tests: decorator createRestControllerDecorator should work for each shortcut decorator 3`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "PATCH",
  },
  "status": 200,
  "text": "PATCH",
}
`;

exports[`Integration Tests: decorator createRestControllerDecorator should work for each shortcut decorator 4`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "PUT",
  },
  "status": 200,
  "text": "PUT",
}
`;

exports[`Integration Tests: decorator createRestControllerDecorator should work for each shortcut decorator 5`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "POST",
  },
  "status": 200,
  "text": "POST",
}
`;

exports[`Integration Tests: decorator createRestControllerDecorator should work for each shortcut decorator 6`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Integration Tests: decorator createRestControllerDecorator should work for methods which call nextFunc() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "\\"GET\\"",
}
`;

exports[`Integration Tests: decorator createRestControllerDecorator should work for more obscure HTTP methods using the httpMethod decorator 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "8",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "PROPFIND",
  },
  "status": 200,
  "text": "PROPFIND",
}
`;

exports[`Integration Tests: decorator createRestControllerDecorator should work with config 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "16",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "\\"config ok !!!!\\"",
}
`;
